File: object-oriented-programming/gcse-concepts/c-sharp/pet_class.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4
Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/


using System;
namespace IsaacCodeSamples
{
    class Pet {
        private string name;
        private string type;
        private string colour;
        private string mood;
        private Boolean sleeping;
        
        
        // Constructor
        public Pet(string givenName, string givenType, string givenColour) {
            name = givenName;
            type = givenType;
            colour = givenColour;
            string[] allMoods = { "playful", "hungry", "sleepy" };
            Random rnd = new Random();
            int index = rnd.Next(allMoods.Length);
            mood = allMoods[index];
            sleeping = false;
        }
        
        
        public string GetName() {
            return name;
        }
        
        
        public new string GetType() {
            return type;
        }
        
        
        public string GetColour() {
            return colour;
        }
        
        
        public string GetMood() {
            return mood;
        }
        
        
        public Boolean IsSleeping() {
            return sleeping;
        }
        
        
        public void Describe() {
            Console.WriteLine($"I am a {mood}, {colour} {type} called {name}");
        }
        
        
        public void SetName(string newName) {
            name = newName;
        }
        
        
        public void SetType(string newType) {
            type = newType;
        }
        
        
        public void SetColour(string newColour) {
            colour = newColour;
        }
        
        
        public void Play() {
            if (sleeping == true) {
                Console.WriteLine("Zzzzzzz. I am sleeping");
            } else if (mood == "hungry") {
                Console.WriteLine("I am too hungry to play");
            } else if (mood == "tired") {
                Console.WriteLine("I am too tired to play");
            } else {
                Console.WriteLine("This is fun, I love playing");            
                string[] allMoods = { "playful", "hungry", "sleepy" };
                Random rnd = new Random();
                int index = rnd.Next(allMoods.Length);
                mood = allMoods[index];
            }
        }
        
        
        public void Feed() {
            if (sleeping == true) {
                Console.WriteLine("Zzzzzzz. I am sleeping");
            }
            if (mood == "tired") {
                Console.WriteLine("I am too sleepy to eat anything now");
            } else if (mood == "playful") {
                Console.WriteLine("I am not hungry - I want to play");
            } else {
                Console.WriteLine("Yum - yum - that tastes great");
                mood = "playful";
            }
        }
        
        
        public void Sleep() {
            if (mood == "playful") {
                Console.WriteLine("I am too playful to sleep");
            } else if (mood == "hungry") {
                Console.WriteLine("I need something to eat before I can go to sleep");
            } else {
                Console.WriteLine("Zzzzzzzz");
                sleeping = true;
            }
        }
        
        
        public void Wake() {
            if (sleeping == false) {
                Console.WriteLine("I wasn't even asleep!");
            } else {
                Console.WriteLine("I am awake and ready to play!");
                sleeping = false;
                mood = "playful";
            }
        }
        
    }
}

File: object-oriented-programming/gcse-concepts/python/pet_class.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4
# https://isaaccomputerscience.org/concepts/prog_oop_concepts

import random

class Pet:
    """Defines a pet in a children's game"""
    def __init__(self, given_name, given_type, given_colour):
        self.name = given_name
        self.type = given_type
        self.colour = given_colour
        all_moods = ["playful", "hungry", "sleepy"]
        self.mood = random.choice(all_moods)
        self.sleeping = False

    def get_name(self):
        return self.name

    def get_type(self):
        return self.type

    def get_colour(self):
        return self.colour

    def get_mood(self):
        return self.mood

    def is_sleeping(self):
        return self.sleeping

    def describe(self):
        print(f"I am a {self.mood}, {self.colour} {self.type} called {self.name}")

    def set_name(self, new_name):
        self.name = new_name

    def set_type(self, new_type):
        self.type = new_type

    def set_colour(self, new_colour):
        self.colour = new_colour

    def play(self):
       if self.sleeping == True:
           print("Zzzzzzz. I am sleeping")
       elif self.mood == "hungry":
           print("I am too hungry to play")
       elif self.mood == "tired":
           print("I am too tired to play")
       else:
           print("This is fun, I love playing")
           all_moods = ["playful", "hungry", "sleepy"]
           self.mood = random.choice(all_moods)

    def feed(self):
        if self.sleeping == True:
           print("Zzzzzzz. I am sleeping")
        if self.mood == "tired":
            print("I am too sleepy to eat anything now")
        elif self.mood == "playful":
            print("I am not hungry - I want to play")
        else:
            print("Yum - yum - that tastes great")
            self.mood = "playful"

    def sleep(self):
        if self.mood == "playful":
            print("I am too playful to sleep")
        elif self.mood == "hungry":
            print("I need something to eat before I can go to sleep")
        else:
            print("Zzzzzzzz")
            self.sleeping = True

    def wake(self):
        if self.sleeping == False:
           print("I wasn't even asleep!")
        else:
            print("I am awake and ready to play!")
            self.sleeping = False
            self.mood = "playful"


